biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java

```diff
@@ -0,0 +1,145 @@
+package aQute.bnd.osgi;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import aQute.bnd.header.Attrs;
+import aQute.bnd.osgi.Descriptors.PackageRef;
+
+public class PermissionGenerator {
+	public static final String	KEY					= "permissions";
+
+	public static final String	PARAM_ALL			= "all";
+	public static final String	PARAM_ADMIN			= "admin";
+	public static final String	PARAM_CAPABILITY	= "capability";
+	public static final String	PARAM_PACKAGES		= "packages";
+	public static final String	PARAM_SERVICES		= "services";
+
+	private final Builder		builder;
+	private final List<String>	args;
+	private final StringBuilder	sb;
+
+	public PermissionGenerator(Builder builder, final String... args) {
+		this.builder = builder;
+
+		if (args.length == 2 && PARAM_ALL.equals(args[1])) {
```

Comment:
In the original code `all` was equivalent to `packages` since each iteration over args would match the first if in the if/else chain. `${permissions; admin; all}` was the same as `${permissions; admin; packages}`.